lms/util/lti.py

```diff
@@ -0,0 +1,30 @@
+from pyramid.httpexceptions import HTTPBadRequest
+
+from lms import models
+
+
+def lti_params_for(request):
+    """
+    Return the LTI params for the given request.
+
+    If the request is an LTI launch request then just return request.params.
+
+    If the request is an OAuth 2.0 redirect request then use the request's
+    `state` param to retrieve the launch params that were previously stashed in
+    the DB and return them.
+
+    :raise HTTPBadRequest: if the request is an OAuth redirect but no
+      associated launch params can be found in the DB
+
+    """
+    if "state" in request.params:
+        # This is an OAuth redirect request, not an LTI launch request.
+        # Retrieve the LTI launch params from the database instead of using
+        # request.params directly.
+        lti_params = models.find_lti_params(request.db, request.params["state"])
+        if lti_params is None:
+            raise HTTPBadRequest("OAuth state was not found")
+        return lti_params
+    else:
+        # This is an LTI launch request.
+        return request.params
```

Comment:
I'm not going to a huge amount of trouble to make sure that what this function sees as an LTI launch req is in fact a launch req, and what it sees as an OAuth redirect req is in fact a redirect req. Partially because I think the whole handling of OAuth redirects should in future be done in a completely different way and this won't be needed at all. And partially because the app needs centralised request params validation code, also intended for future refactoring. So I'm keeping it quick and dirty here.

Reply:
Yeah, my brain had been going down these same lines of inquiry, but I think the hokey-and-dirty check on the request param for now is fine...though I wonder if it would be helpful to abstract it out into a util function or something? In case there are other parts of code who need to ask the question "is this an LTI launch request or an OAuth request?"